client/src/lib/dependency-injection/container.js

```diff
@@ -0,0 +1,235 @@
+import toposort from 'toposort';
+import { compose } from 'redux';
+
+/**
+ * Store of middleware functions for composing dependencies
+ * @type Object
+ */
+const middlewares = {};
+
+const container = {};
+
+/**
+ * A list of allowed priorities that can be specified as metadata
+ * @type array
+ */
+const PRIORITIES = ['before', 'after'];
+/**
+ * When true, DI is blocked
+ * @type {boolean}
+ */
+let initialised = false;
+
+/**
+ * Validates the metadata passed to the injector customisation
+ * @param object
+ */
+const validateMeta = (meta) => {
+  if (typeof meta.name === undefined) {
+    throw new Error('Injector.update() called with no "name" property specified.');
+  }
+  PRIORITIES.forEach(k => {
+    if (
+      meta[k] !== undefined &&
+      (typeof meta[k] !== 'string' && !Array.isArray(meta[k]))
+    ) {
+      throw new Error(`Injector.update() key ${k} must be a string or array`);
+    }
+  });
+};
+
+/**
+ * Wrapper function to enforce frozen state of DI container
+ * @param func
+ * @returns function
+ */
+const protect = (func) => (...params) => {
+  if (initialised) {
+    throw new Error('Cannot mutate DI container after it has been initialised');
+  }
+
+  func(...params);
+};
+
+/**
+ * Creates a display name for a final composed component given all
+ * the names of the mutations that affected it.
+ * e.g. my-module(TextField)
+ * @param original The original registered component
+ * @param module The list of module names that modified the component
+ */
+const createDisplayName = (original, modules) => {
+  const componentName = (original.displayName || original.name || 'Component');
+  const names = [componentName, ...modules];
+
+  return names.reduce((acc, curr) => `${curr}(${acc})`);
+};
+
+/**
+ * Ensures that the priority keys are arrays
+ * @param middleware
+ * @returns object
+ */
+const normaliseMiddleware = (middleware) => {
+  const normalised = { ...middleware };
+  PRIORITIES.forEach(k => {
+    if (!Array.isArray(middleware[k])) {
+      normalised[k] = middleware[k] ? [middleware[k]] : [];
+    } else {
+      normalised[k] = middleware[k];
+    }
+  });
+  return normalised;
+};
+
+/**
+ * Applies a topological sort to the middlewares and resolves
+ * priority declarations
+ * @param middlewareList
+ * @returns {Array}
+ */
+const sortMiddlewares = (middlewareList) => {
+  const graph = [];
+  let sortedMiddlewares = [];
+
+  middlewareList.forEach(entry => {
```

Comment:
So if we apply a HoC in `silverstripe-admin/` on a "base component" (can't think of a use case for this at the moment...), does the `*` allow modules to apply their customisations *before*?